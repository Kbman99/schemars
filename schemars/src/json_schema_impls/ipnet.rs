use crate::gen::SchemaGenerator;
use crate::schema::*;
use crate::JsonSchema;
use ipnet::{IpNet, Ipv4Net, Ipv6Net};

impl JsonSchema for IpNet {
    fn schema_name() -> String {
        "IpNet".to_string()
    }

    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
        SchemaObject {
            metadata: Some(Box::new(Metadata {
                description: Some("Represents a generic network address range. This type can have two variants: v4 and v6.".to_string()),
                examples: vec![
                    serde_json::to_value(IpNet::V4("127.0.0.0/24".parse().unwrap())),
                    serde_json::to_value(IpNet::V6("dead::/7".parse().unwrap())),
                ].into_iter().flatten().collect(),
                ..Default::default()
            })),
            subschemas: Some(Box::new(SubschemaValidation {
                any_of: Some(vec![
                    gen.subschema_for::<Ipv4Net>(),
                    gen.subschema_for::<Ipv6Net>(),
                ]),
                ..Default::default()
            })),
            ..Default::default()
        }
        .into()
    }
}

impl JsonSchema for Ipv4Net {
    fn schema_name() -> String {
        "Ipv4Net".to_string()
    }

    fn json_schema(_: &mut SchemaGenerator) -> Schema {
        SchemaObject {
            metadata: Some(Box::new(Metadata {
                description: Some(
                    "Represents a network address range where the IP addresses are v4 variants".to_string(),
                ),
                examples: vec![serde_json::to_value(
                    "127.0.0.0/24".parse::<Ipv4Net>().unwrap(),
                )]
                .into_iter()
                .flatten()
                .collect(),
                ..Default::default()
            })),
            instance_type: Some(InstanceType::String.into()),
            format: Some("ipv4-cidr".to_string()),
            ..Default::default()
        }
        .into()
    }
}

impl JsonSchema for Ipv6Net {
    fn schema_name() -> String {
        "Ipv6Net".to_string()
    }

    fn json_schema(_: &mut SchemaGenerator) -> Schema {
        SchemaObject {
            metadata: Some(Box::new(Metadata {
                description: Some(
                    "Represents a network address range where the IP addresses are v4 variants".to_string(),
                ),
                examples: vec![serde_json::to_value(
                    "dead::/7".parse::<Ipv6Net>().unwrap(),
                )]
                .into_iter()
                .flatten()
                .collect(),
                ..Default::default()
            })),
            instance_type: Some(InstanceType::String.into()),
            format: Some("ipv6-cidr".to_string()),
            ..Default::default()
        }
        .into()
    }
}
