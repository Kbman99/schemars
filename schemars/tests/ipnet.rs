mod util;
use ipnet::IpNet;
use util::*;

#[test]
fn ipnet() -> TestResult {
    test_default_generated_schema::<IpNet>("ipnet")
}
