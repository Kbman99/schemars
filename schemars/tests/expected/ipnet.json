{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "IpNet",
  "description": "Represents a generic network address range. This type can have two variants: v4 and v6.",
  "examples": [
    "127.0.0.0/24",
    "dead::/7"
  ],
  "anyOf": [
    {
      "$ref": "#/definitions/Ipv4Net"
    },
    {
      "$ref": "#/definitions/Ipv6Net"
    }
  ],
  "definitions": {
    "Ipv4Net": {
      "description": "Represents a network address range where the IP addresses are v4 variants",
      "examples": [
        "127.0.0.0/24"
      ],
      "type": "string",
      "format": "ipv4-cidr"
    },
    "Ipv6Net": {
      "description": "Represents a network address range where the IP addresses are v6 variants",
      "examples": [
        "dead::/7"
      ],
      "type": "string",
      "format": "ipv6-cidr"
    }
  }
}